crates/spidermonkey-embedding-splicer/src/bindgen.rs

```diff
@@ -516,8 +521,32 @@ impl JsBindgen<'_> {
                             );
                         }
 
-                        if resource.is_some() {
-                            uwriteln!(self.src, "}}");
+                        if let Some(ty) = resource {
+                            let lower_camel = &self.resolve.types[ty]
+                                .name
+                                .as_ref()
+                                .unwrap()
+                                .to_lower_camel_case();
+
+                            let prefix = iface_name
+                                .as_deref()
+                                .map(|s| format!("{s}$"))
+                                .unwrap_or(String::new());
+
+                            let resource_symbol = self.intrinsic(Intrinsic::SymbolResourceHandle);
+                            let dispose_symbol = self.intrinsic(Intrinsic::SymbolDispose);
+
+                            uwriteln!(
+                                self.src,
+                                "
+                                    [{dispose_symbol}]() {{
+                                        finalizationRegistry_import${prefix}{lower_camel}.unregister(this);
+                                        $resource_import${prefix}drop${lower_camel}(this[{resource_symbol}]);
```

Comment:
Do we need to separate the borrow case here? Or should we be dropping borrows as well? The finalization registry avoids a double drop, but are we not already explicitly dropping borrows at the call end?

Reply:
I don't currently have a very clear picture of the resource model, but some of my questions are:
* Currently, without this PR, are all borrows dropped via the finalization registry? Does that mean that they are not dropped at the end of the call as well? And if there is a drop of a borrow at the end of the call, if a borrow is dropped by the finalization registry before that end of the call, does that create a double drop or are there double drop protections in this case?
* Then secondly, with this PR, would the same not apply, where calling symbol.dispose might pre-empt the default end-of-call drop that happens?

Also I may well be misinterpreting the model, resources are complex, so let me know if there are any gaps in my interpretation too.

I don't think we should have separate classes, but just that I would have thought that the finalization registry would only be needed for owns and not borrows in general (which is what jco does on the host side).

Perhaps this is more an implementation item separate to this PR though?

Reply:
Well, now I'm confused.  I just pushed a commit that expands the `resource-borrow-import` test to cover a couple of relevant cases, but I can't figure out how to make them fail.  The finalizer function doesn't seem to be run, and yet the host doesn't complain the the borrow handle is not dropped.  Perhaps Jco isn't as picky as it should be about whether the guest drops its borrow handles before returning?

Reply:
Ok, it sounds like this is an existing issue then - let's track this further in https://github.com/bytecodealliance/componentize-js/issues/70 and https://github.com/bytecodealliance/jco/issues/249, and get this PR landed in the mean time.